community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/commands/expressions/ExtractFunction.scala

```diff
@@ -31,9 +31,15 @@ case class ExtractFunction(collection: Expression, id: String, expression: Expre
   with Closure {
   override def compute(value: AnyValue, m: ExecutionContext, state: QueryState): ListValue = {
     val list = makeTraversable(value)
-    VirtualValues.transform(list, new java.util.function.Function[AnyValue, AnyValue]() {
-      override def apply(v1: AnyValue): AnyValue = expression(m.newWith1(id, v1), state)
-    })
+    val extracted = new Array[AnyValue](list.size())
+    val values = list.iterator()
```

Comment:
We could do this without the iterator and using below instead `list.value(i)` it will be slightly better in most cases except for the cases when we have a filtered list.